front/components/markdown/VisualizationBlock.tsx

```diff
@@ -0,0 +1,123 @@
+import type { LightWorkspaceType } from "@dust-tt/types";
+import { useContext, useMemo } from "react";
+import { visit } from "unist-util-visit";
+
+import { VisualizationActionIframe } from "@app/components/assistant/conversation/actions/VisualizationActionIframe";
+import { MarkdownContentContext } from "@app/components/assistant/markdown/MarkdownContentContext";
+import type { CustomRenderers } from "@app/components/assistant/markdown/RenderMessageMarkdown";
+
+const VISUALIZATION_MAGIC_LINE = "{/** visualization-complete */}";
+
+type PositionType = { start: { line: number }; end: { line: number } };
+
+export function VisualizationBlock({
+  position,
+  customRenderer,
+}: {
+  position: PositionType;
+  customRenderer?: CustomRenderers;
+}) {
+  const { content } = useContext(MarkdownContentContext);
+
+  const visualizationRenderer = useMemo(() => {
+    return (
+      customRenderer?.visualization ||
+      (() => (
+        <div className="pb-2 pt-4 font-medium text-red-600">
+          Visualization not available
+        </div>
+      ))
+    );
+  }, [customRenderer]);
+
+  let code = content
+    .split("\n")
+    .slice(position.start.line, position.end.line - 1)
+    .join("\n");
+  let complete = false;
+  if (code.includes(VISUALIZATION_MAGIC_LINE)) {
+    code = code.replace(VISUALIZATION_MAGIC_LINE, "");
+    complete = true;
+  }
+  return visualizationRenderer(code, complete, position.start.line);
+}
+
+export function getVisualizationPlugin(
+  owner: LightWorkspaceType,
+  agentConfigurationId: string,
+  conversationId: string,
+  messageId: string
+) {
+  const VisualizationPlugin = ({ position }: { position: PositionType }) => {
+    return (
+      <VisualizationBlock
+        position={position}
+        customRenderer={{
```

Comment:
Even though it's memoized at call site, I'd memoize the `customRenderer` here as well to avoid re-rendering when position changes.